### PR TITLE
fix(eslint-plugin): [explicit-module-boundary-types] and [explicit-function-return-type] don't report on `as const satisfies`

### LINKED ISSUE #10231: Bug: [explicit-module-boundary-types] `allowDirectConstAssertionInArrowFunctions` should not error when also using `satisfies`
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.6.3&fileType=.ts&code=FAFwngDgpgBAYgewTAvDA3sGMBmSBcMAziAE4CWAdgOYA0WMARgIamGUCuAto1KfQF8A3MGBQAHhASkQMAMYJKJGM1QwAFAEpUAPg2ZseBIQBEcAPKWT9bCzYwAjACYAzIOZF5ikpqEwA9P4wADwAtKEwACoAFuSeAO7SANaeHjAS0HIgUAAmAHSiGdKyCkqyjGpauvoMRqYWVjZMrITObsACKp6lykTMIHE45FCeiAi%2BAUFhETFxMCDRpAjxqZTppEukBUA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6RAD2MbMroFsA9gBMEiWgCNBsJsICG0QrSKlUGANrhsORNGiDokADRbtwKHPjxBAdwAilaIjL4AwoKbJ8AQWTJd%2BJQeAJJM3nq2AGIyLkGe6GD4cIhgAL6mYAC6WumpQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
type Foo = {
  foo: string,
  bar: number,
};

export const a = () => ({
  foo: "FOOO",
  bar: 123,
}as const); // <-- This works as expected.

export const b = () => ({
  foo: "FOOO",
  bar: 123,
} as const satisfies Foo); // <-- This throws an error.
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/explicit-module-boundary-types": [
      "error",
      { "allowDirectConstAssertionInArrowFunctions": true }
    ]
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

This concerns both `explicit-module-boundary-types` and `explicit-function-return-type` rules.

I have a situation where I want to use both `as const` followed by `satisfies`. With the rule currently as written as soon as you add a `satisfies` after an `as const` the rule will error. Leaving off the `satisfies` works as expected. 

In this scenario, the `satisfies Foo` provides extra type safety ensuring that `a satisfies Foo` but still keeping the ability to have explicit type literals for `foo: "FOOO"` and `bar: 123`. 

### Actual Result

The rule should not throw an error when using `as const satisfies Foo`

### Additional Info

_No response_

### PR DESCRIPTION
**Pull request by ronami**
## PR Checklist

- [x] Addresses an existing open issue: fixes #10231
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [Contributing](https://typescript-eslint.io/contributing) were taken

## Overview

This PR resolves #10231 and adjusts both `explicit-module-boundary-types` and `explicit-function-return-type` to not fail on the following when `allowDirectConstAssertionInArrowFunctions` is enabled:

```ts
type Foo = {
  foo: string,
  bar: number,
};

export const a = () => ({
  foo: "FOOO",
  bar: 123,
}as const); // <-- This works as expected.

export const b = () => ({
  foo: "FOOO",
  bar: 123,
} as const satisfies Foo); // <-- This unexpectedly causes the rules to report.
```

### CODE REVIEW

File: packages/eslint-plugin/src/util/explicitReturnTypeUtils.ts [hunk 1]
```diff
@@ -256,11 +253,20 @@ function isValidFunctionExpressionReturnType(
...
+    return false;
+  }
+
+  const { body } = node;
+
+  if (body.type === AST_NODE_TYPES.TSSatisfiesExpression) {
+    return isConstAssertion(body.expression);
+  }
+
+  return isConstAssertion(body);
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/util/explicitReturnTypeUtils.ts:269]: [Style] This is super nitpicky, but I found collapsing down to one statement to be a bit more readable:

```suggestion
  return isConstAssertion(
    body.type === AST_NODE_TYPES.TSSatisfiesExpression ? body.expression : body,
  );
```

No worries at all if you hate this 😄

File: packages/eslint-plugin/tests/rules/explicit-function-return-type.test.ts
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/tests/rules/explicit-function-return-type.test.ts:1]: [Testing] One thing under test per case please (https://typescript-eslint.io/contributing/pull-requests#granular-unit-tests)

**Review by JoshuaKGoldberg — CHANGES_REQUESTED** [review]
Great! Just requesting changes on the tests, not the teeny nitpick.

Alec Baldwin as Jack Donaghy in 30 Rock enthusiastically giving two thumbs up. Caption: "GREAT!"

File: packages/eslint-plugin/src/util/explicitReturnTypeUtils.ts (hunk 1 above)
Comment (ronami) [reply packages/eslint-plugin/src/util/explicitReturnTypeUtils.ts:269]: Not my cup of tea (🍵) but is is more concise 👍

File: packages/eslint-plugin/tests/rules/explicit-function-return-type.test.ts
Comment (ronami) [reply packages/eslint-plugin/tests/rules/explicit-function-return-type.test.ts:1]: I was relying on older tests that checked `as const`, I changed the ones I added to only test a single case each 👍

**Review by JoshuaKGoldberg — DISMISSED** [review]
❤️‍🔥

File: packages/eslint-plugin/src/util/explicitReturnTypeUtils.ts [hunk 2]
```diff
@@ -256,10 +253,17 @@ function isValidFunctionExpressionReturnType(
...
+    node.type !== AST_NODE_TYPES.ArrowFunctionExpression
+  ) {
+    return false;
+  }
+
+  const { body } = node;
+
+  return isConstAssertion(
+    body.type === AST_NODE_TYPES.TSSatisfiesExpression ? body.expression : body,
   );
```
Comment (kirkwaiblinger) [inline packages/eslint-plugin/src/util/explicitReturnTypeUtils.ts:267]: You might wanna use a `while` loop here, like we do in other rules for recursive syntaxes.
```suggestion
  let body = node.body;
  while (body.type === AST_NODE_TYPES.TSSatisfiesExpression) {
    body = body.expression;
  }

  return isConstAssertion(body);
```
Not sure if there's any real world use case but here's a test case 🤷 
```ts
type FooBar = {
  foo: string,
  bar: number,
};

type EmptyObjectType = {}

export const b = () => ({
  foo: "FOOO",
  bar: 123,
} as const satisfies FooBar satisfies EmptyObjectType); 
```

**Review by kirkwaiblinger — DISMISSED** [review]

File: packages/eslint-plugin/src/util/explicitReturnTypeUtils.ts (hunk 2 above)
Comment (ronami) [reply packages/eslint-plugin/src/util/explicitReturnTypeUtils.ts:267]: That's interesting, I didn't consider this edge case, thanks 👍

Comment (ronami) [reply packages/eslint-plugin/src/util/explicitReturnTypeUtils.ts]: I added the necessary tests to cover this.

File: packages/eslint-plugin/tests/rules/explicit-function-return-type.test.ts [hunk 3]
```diff
@@ -468,6 +468,27 @@ const func = (value: number) => x as const satisfies number;
     },
     {
       code: `
+const func = (value: number) => x as const satisfies 10 satisfies number;
```
Comment (kirkwaiblinger) [inline packages/eslint-plugin/tests/rules/explicit-function-return-type.test.ts:471]: fun fact, this code is illegal TS, since `as const` cannot be used on identifiers. Sometimes we intentionally write test cases with illegal TS, but lately we've generally been trying to avoid doing so (or we try to provide a `@ts-expect-error` comment to indicate we're intentionally testing illegal TS). Let's use an alternative here (and in the explicit-module-boundary test case too) that isn't an error since we can easily swap it out with one? See https://github.com/typescript-eslint/typescript-eslint/pull/10315#discussion_r1836146131 for a (surprisingly tedious to craft) example of a legal nested-`satisfies` `as const` pattern.

**Review by kirkwaiblinger — CHANGES_REQUESTED** [review]

File: packages/eslint-plugin/tests/rules/explicit-function-return-type.test.ts (hunk 3 above)
Comment (ronami) [reply packages/eslint-plugin/tests/rules/explicit-function-return-type.test.ts]: Sure, I've updated the tests I added to be valid TypeScript 👍

**Review by kirkwaiblinger — DISMISSED** [review]

**Review by bradzacher — DISMISSED** [review]

**Review by bradzacher — APPROVED** [review]
